### DISCUSSION TITLE
Support ALTER TABLE t ADD CONSTRAINT fk_t_p FOREIGN KEY (p_id) REFERENCES p(id)

### DISCUSSION THREAD

**Discussion by RandomFractals** [duckdb/duckdb#4204] [Ideas]
to enable adding foreign keys and table references after tables are created.

Simple usage scenario:

```
ALTER TABLE child
  ADD CONSTRAINT fk_child_parent
  FOREIGN KEY (parent_id) 
  REFERENCES parent(id);
```

Also, see related ticket #4201 for the foreign keys on/off pragma feature request.

**Comment by Mytherin** [comment]
This is related to issue #57 as well

**Reply by RandomFractals** [reply]
yes. while related, this feature request is more specific to simplify creating tables with foreign key references and importing data from other data sources that have those relationships well defined. 

One example could be importing data from [datapackage](https://github.com/frictionlessdata/datapackage-js) JSON or YAML data files format for tabular data sources import.

See Datapackage YAML JS notebook on observable: https://observablehq.com/@randomfractals/datapackage-yaml?collection=@randomfractals/datapackage

Also, would be nice to have either `pragma foreign_keys = off` as described in #4205 or this feature implemented to simplify tables creation and duckdb initialization for data imports.

**Comment by RandomFractals** [comment]
Better example when this would be used (not quite in duckdb SQL format, but listed in the order of execution for illustration):

```sql
CREATE TABLE Orders (
    OrderID int NOT NULL PRIMARY KEY,
    OrderNumber int NOT NULL,
    PersonID int NOT NULL
);
```

```sql
CREATE TABLE Persons (
    PersonID int NOT NULL PRIMARY KEY,
    LastName varchar(255) NOT NULL,
    FirstName varchar(255),
    Age int
);
```
Using alter table add foreign key constraint after `Orders` table has been created:

```sql
ALTER TABLE Orders
ADD CONSTRAINT FK_PersonOrder
FOREIGN KEY (PersonID) REFERENCES Persons(PersonID);
```

**Comment by pachadotdev** [comment]
I support this. It would be more efficient to copy the tables first, and then add primary and foreign keys.
